### PR TITLE
rd - Save Image Link

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by revatidamle**
Closes #2 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

https://trello.com/c/w8qsYuZL/1805-rails-onboarding-image-sharer-application-task-2

### CODE REVIEW

File: app/assets/javascripts/images.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (joelbandi) [inline app/assets/javascripts/images.js]: lets drop these empty files :)

File: app/assets/stylesheets/images.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Images controller here.
```
Comment (joelbandi) [inline app/assets/stylesheets/images.scss]: this too

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class ImagesController < ApplicationController
+  include ImagesHelper
+  def home
```
Comment (joelbandi) [inline app/controllers/images_controller.rb]: no need for this empty guy to be here

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  end
+  def create
+    image_params =  params.require(:image).permit([:title, :link])
+    url = image_params[:link]
+    valid = validate_url(url)
+    raise ArgumentError.new("Not a valid URL") if not valid
```
Comment (joelbandi) [inline app/controllers/images_controller.rb]: Models are a great place for validators!! It is a really awesome ActiveRecord feature!

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    @image.save
+    redirect_to @image
+  end
+  def show
+    @image = Image.find(params[:id])
+    redirect_to @image.link
+
+  end
+end
+# def index
```
Comment (joelbandi) [inline app/controllers/images_controller.rb]: comments

File: db/migrate/20190221003541_create_images.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class CreateImages < ActiveRecord::Migration[5.2]
```
Comment (joelbandi) [inline db/migrate/20190221003541_create_images.rb]: Awesome!!

File: test/fixtures/images.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (joelbandi) [inline test/fixtures/images.yml]: no need to commit fixtures yet

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
```
Comment (joelbandi) [inline test/controllers/images_controller_test.rb]: missing tests for controllers

File: test/models/image_test.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (joelbandi) [inline test/models/image_test.rb]: missing tests for model

File: app/validators/url_format_validator.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class UrlFormatValidator <ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    valid_url_regex = /^(http|https):\/\/[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(([0-9]{1,5})?\/.*)?$/ix rescue false
```
Comment (grahamotte) [inline app/validators/url_format_validator.rb]: This regex isn't super maintainable without some very thorough tests. I would suggest delegating this parsing to `URI`. 

Also kudos on using the custom validator! nice!

Comment (revatidamle) [reply app/validators/url_format_validator.rb]: Sure. added that!

**Comment by joelbandi** [comment]
@revatidamle looks like rubocop errors

File: app/validators/url_format_validator.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+require 'uri'
+class UrlFormatValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    uri = URI.parse(value)
+    record.errors.add(attribute, 'Invalid URL: ' + value) unless %w[https http].include?(uri.scheme)
+    valid_image_regex = /.\.(png|jpeg|jpg|gif)$/
```
Comment (grahamotte) [inline app/validators/url_format_validator.rb]: This is fine for this project but not all img urls will include the `.ext`, so this isn't strictly a necessary validation

Comment (revatidamle) [reply app/validators/url_format_validator.rb]: Will keep that in mind

**Pull request merged**